apps/api/migrations/preferences-uniqueness/preferences-uniqueness-migration.ts

```diff
@@ -0,0 +1,146 @@
+
+  logger.info('start migration - preferences uniqueness');
+
+  const promiseTypes = [
+    PreferencesTypeEnum.SUBSCRIBER_GLOBAL,
+    PreferencesTypeEnum.SUBSCRIBER_WORKFLOW,
+    PreferencesTypeEnum.USER_WORKFLOW,
+    PreferencesTypeEnum.WORKFLOW_RESOURCE,
+  ];
+
+  const promises: Promise<void>[] = [];
+  // Subscriber global preferences
+  promises.push(
+    deletePreferenceDuplicates({
+      preferencesRepository,
+      logger,
+      type: PreferencesTypeEnum.SUBSCRIBER_GLOBAL,
+      sort: { _environmentId: 1, _subscriberId: 1 },
+      groupId: { _environmentId: '$_environmentId', _subscriberId: '$_subscriberId' },
+    })
+  );
+
+  // Subscriber workflow preferences
+  promises.push(
+    deletePreferenceDuplicates({
+      preferencesRepository,
+      logger,
+      type: PreferencesTypeEnum.SUBSCRIBER_WORKFLOW,
+      sort: { _environmentId: 1, _subscriberId: 1, _templateId: 1 },
+      groupId: { _environmentId: '$_environmentId', _subscriberId: '$_subscriberId', _templateId: '$_templateId' },
+    })
+  );
+
+  // User workflow preferences
+  promises.push(
+    deletePreferenceDuplicates({
+      preferencesRepository,
+      logger,
+      type: PreferencesTypeEnum.USER_WORKFLOW,
+      sort: { _environmentId: 1, _templateId: 1 },
+      groupId: { _environmentId: '$_environmentId', _templateId: '$_templateId' },
+    })
+  );
+
+  // Workflow resource preferences
+  promises.push(
+    deletePreferenceDuplicates({
+      preferencesRepository,
+      logger,
+      type: PreferencesTypeEnum.WORKFLOW_RESOURCE,
+      sort: { _environmentId: 1, _templateId: 1 },
+      groupId: { _environmentId: '$_environmentId', _templateId: '$_templateId' },
+    })
+  );
+
+  await Promise.allSettled(promises).then((results) => {
+    for (const result of results) {
+      if (result.status === 'rejected') {
+        const index = results.indexOf(result);
+        const promiseType = promiseTypes[index];
+
+        logger.error('error deleting %s preferences duplicates: %s', promiseType, result.reason);
+      }
+    }
+  });
+
+  logger.info('end migration');
+  await app.close();
+}
+
+async function deletePreferenceDuplicates({
+  preferencesRepository,
+  logger,
+  type,
+  sort,
+  groupId,
+}: {
+  preferencesRepository: PreferencesRepository;
+  logger: PinoLogger;
+  type: PreferencesTypeEnum;
+  sort: Record<string, 1 | Expression.Meta | -1>;
+  groupId: Record<string, string>;
+}) {
+  logger.info('deleting %s preferences duplicates', type);
+
+  const cursor = await preferencesRepository._model.aggregate<{
+    ids: { id: string; updatedAt: Date; _environmentId: string }[];
+  }>(
+    [
+      { $match: { type } },
+      { $sort: sort },
+      {
+        $group: {
+          _id: groupId,
+          ids: { $push: { id: '$_id', updatedAt: '$updatedAt', _environmentId: '$_environmentId' } },
+          count: { $sum: 1 },
+        },
+      },
+      { $match: { count: { $gt: 1 } } },
```

Comment:
1. find by preference type
2. sort and laverage index for sorting
3. group by fields that define "uniqueness", push ids, and count
4. match only with count > 1